File: MichiMark/Domain/Model/MarkLinkDomain.swift
import Foundation

struct MarkLinkDomain: Equatable, Sendable {
    let id: MarkLinkID
    var markLinkSeq: Int///表示順
    var markLinkType: MarkOrLink/// mark / link（初期値は .mark 想定）
    var markLinkDate: Date/// 任意の日付に設定可能
    var markLinkName: String?/// 入力必須（Stateでは空欄許可）
    var members: [MemberDomain]///メンバー（空欄列可）
    var meterValue: Int?/// 累積メーター（km）: Mark 用
    var distanceValue: Int?/// 区間距離（km）: Link 用
    var actions: [ActionDomain]/// 行動（空配列可）
    var memo: String?/// 入力必須（Stateでは空欄許可）

    /// 給油フラグ（初期値は false 想定）
    /// isFuel == false の場合、ガソリン関連項目は nil を想定（制約は Adapter/Reducer で担保）
    var isFuel: Bool
    var pricePerGas: Int?/// 単位: 1円/L
    var gasQuantity: Int?/// 単位: 0.1L（10倍値）
    var gasPrice: Int?/// 単位: 1円

    var isDeleted: Bool/// 論理削除
    let createdAt: Date/// 登録日（初回のみ設定）
    var updatedAt: Date/// 更新日（保存時更新）
    init(
        id: MarkLinkID,
        markLinkSeq: Int,
        markLinkType: MarkOrLink = .mark,
        markLinkDate: Date,
        markLinkName: String? = nil,
        members: [MemberDomain] = [],
        meterValue: Int? = nil,
        distanceValue: Int? = nil,
        actions: [ActionDomain] = [],
        memo: String? = nil,
        isFuel: Bool = false,
        pricePerGas: Int? = nil,
        gasQuantity: Int? = nil,
        gasPrice: Int? = nil,
        isDeleted: Bool = false,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.markLinkSeq = markLinkSeq
        self.markLinkType = markLinkType
        self.markLinkDate = markLinkDate
        self.markLinkName = markLinkName
        self.members = members
        self.meterValue = meterValue
        self.distanceValue = distanceValue
        self.actions = actions
        self.memo = memo
        self.isFuel = isFuel
        self.pricePerGas = pricePerGas
        self.gasQuantity = gasQuantity
        self.gasPrice = gasPrice
        self.isDeleted = isDeleted
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

File: flutter/lib/domain/transaction/mark_link/mark_link_domain.dart
import 'package:equatable/equatable.dart';
import '../../master/member/member_domain.dart';
import '../../master/action/action_domain.dart';
import 'mark_or_link.dart';

class MarkLinkDomain extends Equatable {
  final String id;

  /// 表示順
  final int markLinkSeq;

  /// マーク（地点）またはリンク（経路）の区分
  final MarkOrLink markLinkType;

  /// 記録日時
  final DateTime markLinkDate;

  /// 名称（任意）
  final String? markLinkName;

  /// 参加メンバー
  final List<MemberDomain> members;

  /// 累積メーター（単位: km）。Mark用
  final int? meterValue;

  /// 区間距離（単位: km）。Link用
  final int? distanceValue;

  /// 行動リスト
  final List<ActionDomain> actions;

  /// メモ（任意）
  final String? memo;

  /// 給油フラグ（false の場合、給油関連フィールドは null を想定）
  final bool isFuel;

  /// ガソリン単価（単位: 1円/L）
  final int? pricePerGas;

  /// 給油量（単位: 0.1L の10倍整数値。例: 30.5L → 305）
  final int? gasQuantity;

  /// 給油金額（単位: 1円）
  final int? gasPrice;

  /// 論理削除フラグ
  final bool isDeleted;

  /// 登録日（初回のみ設定）
  final DateTime createdAt;

  /// 更新日（保存時更新）
  final DateTime updatedAt;

  /// ガソリン支払者。isFuel == true のとき意味を持つ
  final MemberDomain? gasPayer;

  const MarkLinkDomain({
    required this.id,
    required this.markLinkSeq,
    this.markLinkType = MarkOrLink.mark,
    required this.markLinkDate,
    this.markLinkName,
    this.members = const [],
    this.meterValue,
    this.distanceValue,
    this.actions = const [],
    this.memo,
    this.isFuel = false,
    this.pricePerGas,
    this.gasQuantity,
    this.gasPrice,
    this.isDeleted = false,
    required this.createdAt,
    required this.updatedAt,
    this.gasPayer,
  });

  MarkLinkDomain copyWith({
    String? id,
    int? markLinkSeq,
    MarkOrLink? markLinkType,
    DateTime? markLinkDate,
    String? markLinkName,
    List<MemberDomain>? members,
    int? meterValue,
    int? distanceValue,
    List<ActionDomain>? actions,
    String? memo,
    bool? isFuel,
    int? pricePerGas,
    int? gasQuantity,
    int? gasPrice,
    bool? isDeleted,
    DateTime? createdAt,
    DateTime? updatedAt,
    MemberDomain? gasPayer,
  }) {
    return MarkLinkDomain(
      id: id ?? this.id,
      markLinkSeq: markLinkSeq ?? this.markLinkSeq,
      markLinkType: markLinkType ?? this.markLinkType,
      markLinkDate: markLinkDate ?? this.markLinkDate,
      markLinkName: markLinkName ?? this.markLinkName,
      members: members ?? this.members,
      meterValue: meterValue ?? this.meterValue,
      distanceValue: distanceValue ?? this.distanceValue,
      actions: actions ?? this.actions,
      memo: memo ?? this.memo,
      isFuel: isFuel ?? this.isFuel,
      pricePerGas: pricePerGas ?? this.pricePerGas,
      gasQuantity: gasQuantity ?? this.gasQuantity,
      gasPrice: gasPrice ?? this.gasPrice,
      isDeleted: isDeleted ?? this.isDeleted,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      gasPayer: gasPayer ?? this.gasPayer,
    );
  }

  @override
  List<Object?> get props => [
        id,
        markLinkSeq,
        markLinkType,
        markLinkDate,
        markLinkName,
        members,
        meterValue,
        distanceValue,
        actions,
        memo,
        isFuel,
        pricePerGas,
        gasQuantity,
        gasPrice,
        isDeleted,
        createdAt,
        updatedAt,
        gasPayer,
      ];
}
